[Product Selector] Analytics events update

samiuelson opened the pull request:
### [Product Selector] Analytics events update

Closes: #8624 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
1. Added `product_count` property to `ORDER_PRODUCT_ADD` event
2. Enabled tracking Product Selector actions: 
* product/variation item selected action
* product/variation item unselected action
* "Done" button taps
3. Added new events for tracking the above actions during Order creation flow:
* `ORDER_CREATION_PRODUCT_SELECTOR_ITEM_SELECTED`
* `ORDER_CREATION_PRODUCT_SELECTOR_ITEM_UNSELECTED`
* `ORDER_CREATION_PRODUCT_SELECTOR_CONFIRM_BUTTON_TAPPED`

### Testing instructions
1. Create new order 
2. Open products picker (+ Add products)
3. Select, unselect items and tap the confirmation button at the bottom of the product selector screen

Verify in logcat that the following events are tracked: 
<img width="2042" alt="Screenshot 2023-03-28 at 15 29 00" src="https://user-images.githubusercontent.com/4527432/228252662-794de70e-06b6-47d3-8cfe-7dab0e3b82cc.png">

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Track product selector events

samiuelson committed:
Add product_count prop to ORDER_PRODUCT_ADD event

samiuelson committed:
Update RELEASE-NOTES.txt

samiuelson committed:
Update unit tests

samiuelson committed:
Update unit tests

samiuelson committed:
Fix lint

kidinov reviewed (commented):
I haven't tested it yet properly, but I'd advise taking a bit different approach here: you may not pass the analytics event, but the enum that determines the flow (order creation or coupons). Depending on the flow in the respective VM, you can select what events to use. Look into `nav_graph_payment_flow.xml`  for inspiration. Following this approach, I think you can do the same with 1/3 of the changes, and having knowledge of the flow inside ProductSelector potentially may have used not only for analytics but for different tunings as well 

Also, I think if you declare parameters on the fragment (receiver)  side where you want to go, then there is no need to declare them on all the actions - they will be generated anyway (if not nullable)

samiuelson committed:
Revert "Track product selector events"

This reverts commit 77ffa741f64285effe90bc6a6c8c4ceb7df87466.

samiuelson committed:
Track product selector events

samiuelson committed:
Merge branch 'trunk' into 8624-analytics-events-for-product-selector

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductNavigationTarget.kt

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/selector/VariationSelectorViewModel.kt:
```diff
@@ -132,12 +134,26 @@ class VariationSelectorViewModel @Inject constructor(
     }
 
+    private fun trackVariationSelected() {
+        navArgs.productSelectorFlow.itemSelectedAnalyticsEvent?.let {
```
If you would do something like that here:

```
        when (navArgs.productSelectorFlow) {
            is ProductSelectorFlow.OrderCreation -> {
                tracker.track(ORDER_CREATION_PRODUCT_SELECTOR_ITEM_SELECTED)
            }
            is ProductSelectorFlow.CouponEdition -> {
                tracker.track(
                    ...
                )
            }
        }.exhaustive
```
(flows don't contain the events at all)

you'll have
* Compiler check if a new flow added
* Less likely to make an error passing null as an event because there are no nullable variables at all
* Easier to test this code - all related to tracking in one place

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/analytics/AnalyticsEvent.kt:
```diff
@@ -255,6 +255,9 @@ enum class AnalyticsEvent(val siteless: Boolean = false) {
     ORDER_PRODUCT_REMOVE,
     ORDER_FEE_REMOVE,
     ORDER_SHIPPING_METHOD_REMOVE,
+    ORDER_CREATION_PRODUCT_SELECTOR_ITEM_SELECTED,
```
Clear selection button won't be tracked?

samiuelson replied:
Thanks for the suggestion. It's added.

kidinov commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/products/selector/ProductSelectorViewModelTest.kt:
```diff
@@ -38,23 +40,17 @@ internal class ProductSelectorViewModelTest : BaseUnitTest() {
     }
     private val variationSelectorRepository: VariationSelectorRepository = mock()
     private val resourceProvider: ResourceProvider = mock()
+    private val tracker: AnalyticsTrackerWrapper = mock()
```
There are not test on the changes?

samiuelson replied:
Unit tests are added. Thanks for the suggestion.

kidinov commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/creation/UnifiedOrderEditViewModelTest.kt:
```diff
@@ -118,7 +118,10 @@ abstract class UnifiedOrderEditViewModelTest : BaseUnitTest() {
-            mapOf(AnalyticsTracker.KEY_FLOW to tracksFlow),
+            mapOf(
+                AnalyticsTracker.KEY_FLOW to tracksFlow,
+                AnalyticsTracker.KEY_PRODUCT_COUNT to 1
```
Maybe add a test with another count too?

kidinov reviewed (commented):
@samiuelson Thank you for the changes! It looks good, but I left a couple of questions, just to be sure that all this expected

samiuelson committed:
Refactor way of passing analytics events

samiuelson committed:
Update unit test

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/selector/VariationSelectorViewModel.kt:
```diff
@@ -129,6 +133,20 @@ class VariationSelectorViewModel @Inject constructor(
+                )
+            }
+            else -> {
+                // no-op
```
Having else with no-op defeats the purpose of having restricted class hierarchies of the enum. if another flow will be added compiler happily will compile this code and we will miss a place where we should add tracking

To utilize that we might need to have it like that:

```
ProductSelectorFlow.CouponEdition -> {}
            ProductSelectorFlow.Undefined -> {}
        }.exhaustive
```

samiuelson replied:
Good point, fixed with 63f5faf.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/selector/VariationSelectorViewModel.kt:
```diff
@@ -127,17 +133,59 @@ class VariationSelectorViewModel @Inject constructor(
+        }
+    }
+
+    private fun trackVariationUnselected() {
```
it looks like that that these tracking methods are the same across both fragments. Would it make sense to extract them in some `ProductSelectionTracker`?

samiuelson replied:
Done with 63f5faf.

samiuelson committed:
Refactor product selector events tracking

- Extracted duplicated code to `ProductSelectorTracker` class
- Added analytics event for "Clear selection" button taps
- Added unit tests for analytics events

samiuelson committed:
Add unit tests for variation selector analytics

samiuelson committed:
Add clear selection event tracking unit test

samiuelson commented:
Thanks for the review, @kidinov! I've addressed your comments and will appreciate another round of review 🙌🏼

kidinov reviewed (approved):
Thanks! Looks good!

Don't forget to register new events and sync them with iOS if not yet
